csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
This is how System.Collection.List<T> is implementing AddRange (actually InsertRange) yet in a more complex way.

Depending on implementation of ICollection the copy mehtod relys on Array.Copy internally. Maybe this could be improved by checking for ICollection<T> instead of ICollection. I do not see any issue on this. IMHO RepeatedField shall not take care of what is contained in there, if the user puts null values in repeatefield, the user has to deal with it.
Actually only this "I do not want null values in repeatedfield" constraint is preventing from using ICollection.CopyTo
. Calling Add(T item) in a loop is not performant. Why shall we prevent null values in RepeatedField ?

As a background, we are using streaming large array of floats using GRPC and found repeatedfield to be a serious bottleneck.